File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Classes/ScoutCard.java
package com.alphadevelopmentsolutions.frcscout.Classes;

import com.alphadevelopmentsolutions.frcscout.Enums.StartingPiece;
import com.alphadevelopmentsolutions.frcscout.Enums.StartingPosition;

import java.text.SimpleDateFormat;
import java.util.Date;

public class ScoutCard
{

    public static final String TABLE_NAME = "scout_cards";
    public static final String COLUMN_NAME_ID = "Id";
    public static final String COLUMN_NAME_MATCH_ID = "MatchId";
    public static final String COLUMN_NAME_TEAM_ID = "TeamId";
    public static final String COLUMN_NAME_EVENT_ID = "EventId";
    public static final String COLUMN_NAME_ALLIANCE_COLOR = "AllianceColor";
    public static final String COLUMN_NAME_COMPLETED_BY = "CompletedBy";

    public static final String COLUMN_NAME_PRE_GAME_STARTING_LEVEL = "PreGameStartingLevel";
    public static final String COLUMN_NAME_PRE_GAME_STARTING_POSITION = "PreGameStartingPosition";
    public static final String COLUMN_NAME_PRE_GAME_STARTING_PIECE = "PreGameStartingPiece";

    public static final String COLUMN_NAME_AUTONOMOUS_EXIT_HABITAT = "AutonomousExitHabitat";
    public static final String COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_PICKED_UP = "AutonomousHatchPanelsPickedUp";
    public static final String COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_SECURED = "AutonomousHatchPanelsSecured";
    public static final String COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_SECURED_ATTEMPTS = "AutonomousHatchPanelsSecuredAttempts";
    public static final String COLUMN_NAME_AUTONOMOUS_CARGO_PICKED_UP = "AutonomousCargoPickedUp";
    public static final String COLUMN_NAME_AUTONOMOUS_CARGO_STORED = "AutonomousCargoStored";
    public static final String COLUMN_NAME_AUTONOMOUS_CARGO_STORED_ATTEMPTS = "AutonomousCargoStoredAttempts";

    public static final String COLUMN_NAME_TELEOP_HATCH_PANELS_PICKED_UP = "TeleopHatchPanelsPickedUp";
    public static final String COLUMN_NAME_TELEOP_HATCH_PANELS_SECURED = "TeleopHatchPanelsSecured";
    public static final String COLUMN_NAME_TELEOP_HATCH_PANELS_SECURED_ATTEMPTS = "TeleopHatchPanelsSecuredAttempts";
    public static final String COLUMN_NAME_TELEOP_CARGO_PICKED_UP = "TeleopCargoPickedUp";
    public static final String COLUMN_NAME_TELEOP_CARGO_STORED = "TeleopCargoStored";
    public static final String COLUMN_NAME_TELEOP_CARGO_STORED_ATTEMPTS = "TeleopCargoStoredAttempts";

    public static final String COLUMN_NAME_END_GAME_RETURNED_TO_HABITAT = "EndGameReturnedToHabitat";
    public static final String COLUMN_NAME_END_GAME_RETURNED_TO_HABITAT_ATTEMPTS = "EndGameReturnedToHabitatAttempts";

    public static final String COLUMN_NAME_DEFENSE_RATING = "DefenseRating";
    public static final String COLUMN_NAME_OFFENSE_RATING = "OffenseRating";
    public static final String COLUMN_NAME_DRIVE_RATING = "DriveRating";
    public static final String COLUMN_NAME_NOTES = "Notes";

    public static final String COLUMN_NAME_COMPLETED_DATE = "CompletedDate";
    public static final String COLUMN_NAME_IS_DRAFT = "IsDraft";

    private int id;
    private String matchId;
    private int teamId;
    private String eventId;
    private String allianceColor;
    private String completedBy;

    private int preGameStartingLevel;
    private StartingPosition preGameStartingPosition;
    private StartingPiece preGameStartingPiece;

    private boolean autonomousExitHabitat;
    private int autonomousHatchPanelsPickedUp;
    private int autonomousHatchPanelsSecuredAttempts;
    private int autonomousHatchPanelsSecured;
    private int autonomousCargoPickedUp;
    private int autonomousCargoStoredAttempts;
    private int autonomousCargoStored;
    
    private int teleopHatchPanelsPickedUp;
    private int teleopHatchPanelsSecuredAttempts;
    private int teleopHatchPanelsSecured;
    private int teleopCargoPickedUp;
    private int teleopCargoStoredAttempts;
    private int teleopCargoStored;

    private int endGameReturnedToHabitat;
    private int endGameReturnedToHabitatAttempts;
    
    private int defenseRating;
    private int offenseRating;
    private int driveRating;
    private String notes;
    private Date completedDate;
    private boolean isDraft;

    public ScoutCard(
            int id,
            String matchId,
            int teamId,
            String eventId,
            String allianceColor,
            String completedBy,

            int preGameStartingLevel,
            StartingPosition preGameStartingPosition,
            StartingPiece preGameStartingPiece,

            boolean autonomousExitHabitat,
            int autonomousHatchPanelsPickedUp,
            int autonomousHatchPanelsSecuredAttempts,
            int autonomousHatchPanelsSecured,
            int autonomousCargoPickedUp,
            int autonomousCargoStoredAttempts,
            int autonomousCargoStored,

            int teleopHatchPanelsPickedUp,
            int teleopHatchPanelsSecuredAttempts,
            int teleopHatchPanelsSecured,
            int teleopCargoPickedUp,
            int teleopCargoStoredAttempts,
            int teleopCargoStored,

            int endGameReturnedToHabitat,
            int endGameReturnedToHabitatAttempts,

            int defenseRating,
            int offenseRating,
            int driveRating,
            String notes,
            Date completedDate,
            boolean isDraft)
    {
        this.id = id;
        this.matchId = matchId;
        this.teamId = teamId;
        this.eventId = eventId;
        this.allianceColor = allianceColor;
        this.completedBy = completedBy;

        this.preGameStartingLevel = preGameStartingLevel;
        this.preGameStartingPosition = preGameStartingPosition;
        this.preGameStartingPiece = preGameStartingPiece;

        this.autonomousExitHabitat = autonomousExitHabitat;
        this.autonomousHatchPanelsPickedUp = autonomousHatchPanelsPickedUp;
        this.autonomousHatchPanelsSecuredAttempts = autonomousHatchPanelsSecuredAttempts;
        this.autonomousHatchPanelsSecured = autonomousHatchPanelsSecured;
        this.autonomousCargoPickedUp = autonomousCargoPickedUp;
        this.autonomousCargoStoredAttempts = autonomousCargoStoredAttempts;
        this.autonomousCargoStored = autonomousCargoStored;

        this.teleopHatchPanelsPickedUp = teleopHatchPanelsPickedUp;
        this.teleopHatchPanelsSecuredAttempts = teleopHatchPanelsSecuredAttempts;
        this.teleopHatchPanelsSecured = teleopHatchPanelsSecured;
        this.teleopCargoPickedUp = teleopCargoPickedUp;
        this.teleopCargoStoredAttempts = teleopCargoStoredAttempts;
        this.teleopCargoStored = teleopCargoStored;

        this.endGameReturnedToHabitat = endGameReturnedToHabitat;
        this.endGameReturnedToHabitatAttempts = endGameReturnedToHabitatAttempts;

        this.defenseRating = defenseRating;
        this.offenseRating = offenseRating;
        this.driveRating = driveRating;
        this.notes = notes;
        this.completedDate = completedDate;
        this.isDraft = isDraft;
    }

    /**
     * Used for loading
     * @param id to load
     */
    public ScoutCard(int id)
    {
        this.id = id;
    }

    //region Getters


    public int getId()
    {
        return id;
    }

    public String  getMatchId()
    {
        return matchId;
    }

    public int getTeamId()
    {
        return teamId;
    }

    public String getEventId()
    {
        return eventId;
    }

    public String getAllianceColor()
    {
        return allianceColor;
    }

    public String getCompletedBy()
    {
        return completedBy;
    }

    public int getPreGameStartingLevel()
    {
        return preGameStartingLevel;
    }

    public StartingPosition getPreGameStartingPosition()
    {
        return preGameStartingPosition;
    }

    public StartingPiece getPreGameStartingPiece()
    {
        return preGameStartingPiece;
    }

    public boolean getAutonomousExitHabitat()
    {
        return autonomousExitHabitat;
    }

    public int getAutonomousHatchPanelsPickedUp()
    {
        return autonomousHatchPanelsPickedUp;
    }

    public int getAutonomousHatchPanelsSecuredAttempts()
    {
        return autonomousHatchPanelsSecuredAttempts;
    }

    public int getAutonomousHatchPanelsSecured()
    {
        return autonomousHatchPanelsSecured;
    }

    public int getAutonomousCargoPickedUp()
    {
        return autonomousCargoPickedUp;
    }

    public int getAutonomousCargoStoredAttempts()
    {
        return autonomousCargoStoredAttempts;
    }

    public int getAutonomousCargoStored()
    {
        return autonomousCargoStored;
    }

    public int getTeleopHatchPanelsPickedUp()
    {
        return teleopHatchPanelsPickedUp;
    }

    public int getTeleopHatchPanelsSecuredAttempts()
    {
        return teleopHatchPanelsSecuredAttempts;
    }

    public int getTeleopHatchPanelsSecured()
    {
        return teleopHatchPanelsSecured;
    }

    public int getTeleopCargoPickedUp()
    {
        return teleopCargoPickedUp;
    }

    public int getTeleopCargoStoredAttempts()
    {
        return teleopCargoStoredAttempts;
    }

    public int getTeleopCargoStored()
    {
        return teleopCargoStored;
    }

    public int getEndGameReturnedToHabitat()
    {
        return endGameReturnedToHabitat;
    }

    public int getEndGameReturnedToHabitatAttempts()
    {
        return endGameReturnedToHabitatAttempts;
    }

    public int getDefenseRating()
    {
        return defenseRating;
    }

    public int getOffenseRating()
    {
        return offenseRating;
    }

    public int getDriveRating()
    {
        return driveRating;
    }

    public String getNotes()
    {
        return notes;
    }

    public Date getCompletedDate()
    {
        return completedDate;
    }

    public boolean isDraft()
    {
        return isDraft;
    }

    /**
     * Gets the completed date formated for MySQL timestamp
     * @return MySQL time stamp formatted date
     */
    public String getCompletedDateForSQL()
    {
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd H:mm:ss");

        return simpleDateFormat.format(completedDate);
    }

    //endregion

    //region Setters

    public void setId(int id)
    {
        this.id = id;
    }

    public void setMatchId(String matchId)
    {
        this.matchId = matchId;
    }

    public void setTeamId(int teamId)
    {
        this.teamId = teamId;
    }

    public void setEventId(String eventId)
    {
        this.eventId = eventId;
    }

    public void setAllianceColor(String allianceColor)
    {
        this.allianceColor = allianceColor;
    }

    public void setCompletedBy(String completedBy)
    {
        this.completedBy = completedBy;
    }

    public void setPreGameStartingLevel(int preGameStartingLevel)
    {
        this.preGameStartingLevel = preGameStartingLevel;
    }

    public void setPreGameStartingPosition(StartingPosition preGameStartingPosition)
    {
        this.preGameStartingPosition = preGameStartingPosition;
    }

    public void setPreGameStartingPiece(StartingPiece preGameStartingPiece)
    {
        this.preGameStartingPiece = preGameStartingPiece;
    }

    public void setAutonomousExitHabitat(boolean autonomousExitHabitat)
    {
        this.autonomousExitHabitat = autonomousExitHabitat;
    }

    public void setAutonomousHatchPanelsPickedUp(int autonomousHatchPanelsPickedUp)
    {
        this.autonomousHatchPanelsPickedUp = autonomousHatchPanelsPickedUp;
    }

    public void setAutonomousHatchPanelsSecuredAttempts(int autonomousHatchPanelsSecuredAttempts)
    {
        this.autonomousHatchPanelsSecuredAttempts = autonomousHatchPanelsSecuredAttempts;
    }

    public void setAutonomousHatchPanelsSecured(int autonomousHatchPanelsSecured)
    {
        this.autonomousHatchPanelsSecured = autonomousHatchPanelsSecured;
    }

    public void setAutonomousCargoPickedUp(int autonomousCargoPickedUp)
    {
        this.autonomousCargoPickedUp = autonomousCargoPickedUp;
    }

    public void setAutonomousCargoStoredAttempts(int autonomousCargoStoredAttempts)
    {
        this.autonomousCargoStoredAttempts = autonomousCargoStoredAttempts;
    }

    public void setAutonomousCargoStored(int autonomousCargoStored)
    {
        this.autonomousCargoStored = autonomousCargoStored;
    }

    public void setTeleopHatchPanelsPickedUp(int teleopHatchPanelsPickedUp)
    {
        this.teleopHatchPanelsPickedUp = teleopHatchPanelsPickedUp;
    }

    public void setTeleopHatchPanelsSecuredAttempts(int teleopHatchPanelsSecuredAttempts)
    {
        this.teleopHatchPanelsSecuredAttempts = teleopHatchPanelsSecuredAttempts;
    }

    public void setTeleopHatchPanelsSecured(int teleopHatchPanelsSecured)
    {
        this.teleopHatchPanelsSecured = teleopHatchPanelsSecured;
    }

    public void setTeleopCargoPickedUp(int teleopCargoPickedUp)
    {
        this.teleopCargoPickedUp = teleopCargoPickedUp;
    }

    public void setTeleopCargoStoredAttempts(int teleopCargoStoredAttempts)
    {
        this.teleopCargoStoredAttempts = teleopCargoStoredAttempts;
    }

    public void setTeleopCargoStored(int teleopCargoStored)
    {
        this.teleopCargoStored = teleopCargoStored;
    }

    public void setEndGameReturnedToHabitat(int endGameReturnedToHabitat)
    {
        this.endGameReturnedToHabitat = endGameReturnedToHabitat;
    }

    public void setEndGameReturnedToHabitatAttempts(int endGameReturnedToHabitatAttempts)
    {
        this.endGameReturnedToHabitatAttempts = endGameReturnedToHabitatAttempts;
    }

    public void setDefenseRating(int defenseRating)
    {
        this.defenseRating = defenseRating;
    }

    public void setOffenseRating(int offenseRating)
    {
        this.offenseRating = offenseRating;
    }

    public void setDriveRating(int driveRating)
    {
        this.driveRating = driveRating;
    }

    public void setNotes(String notes)
    {
        this.notes = notes;
    }

    public void setCompletedDate(Date completedDate)
    {
        this.completedDate = completedDate;
    }

    public void setDraft(boolean draft)
    {
        isDraft = draft;
    }


    //endregion

    //region Load, Save & Delete

    /**
     * Loads the ScoutCard from the database and populates all values
     * @param database used for interacting with the SQLITE db
     * @return boolean if successful
     */
    public boolean load(Database database)
    {
        //try to open the DB if it is not open
        if(!database.isOpen()) database.open();

        if(database.isOpen())
        {
            ScoutCard scoutCard = database.getScoutCard(this);


            if (scoutCard != null)
            {
                setMatchId(scoutCard.getMatchId());
                setTeamId(scoutCard.getTeamId());
                setEventId(scoutCard.getEventId());
                setAllianceColor(scoutCard.getAllianceColor());
                setCompletedBy(scoutCard.getCompletedBy());

                setPreGameStartingLevel(scoutCard.getPreGameStartingLevel());
                setPreGameStartingPosition(scoutCard.getPreGameStartingPosition());
                setPreGameStartingPiece(scoutCard.getPreGameStartingPiece());

                setAutonomousExitHabitat(scoutCard.getAutonomousExitHabitat());
                setAutonomousHatchPanelsPickedUp(scoutCard.getAutonomousHatchPanelsPickedUp());
                setAutonomousHatchPanelsSecuredAttempts(scoutCard.getAutonomousHatchPanelsSecuredAttempts());
                setAutonomousHatchPanelsSecured(scoutCard.getAutonomousHatchPanelsSecured());
                setAutonomousCargoPickedUp(scoutCard.getAutonomousCargoPickedUp());
                setAutonomousCargoStoredAttempts(scoutCard.getAutonomousCargoStoredAttempts());
                setAutonomousCargoStored(scoutCard.getAutonomousCargoStored());

                setTeleopHatchPanelsPickedUp(scoutCard.getTeleopHatchPanelsPickedUp());
                setTeleopHatchPanelsSecuredAttempts(scoutCard.getTeleopHatchPanelsSecuredAttempts());
                setTeleopHatchPanelsSecured(scoutCard.getTeleopHatchPanelsSecured());
                setTeleopCargoPickedUp(scoutCard.getTeleopCargoPickedUp());
                setTeleopCargoStoredAttempts(scoutCard.getTeleopCargoStoredAttempts());
                setTeleopCargoStored(scoutCard.getTeleopCargoStored());

                setEndGameReturnedToHabitat(scoutCard.getEndGameReturnedToHabitat());
                setEndGameReturnedToHabitatAttempts(scoutCard.getEndGameReturnedToHabitatAttempts());

                setDefenseRating(scoutCard.getDefenseRating());
                setOffenseRating(scoutCard.getOffenseRating());
                setDriveRating(scoutCard.getDriveRating());
                setNotes(scoutCard.getNotes());
                setCompletedDate(scoutCard.getCompletedDate());
                setDraft(scoutCard.isDraft());
                return true;
            }
        }

        return false;
    }

    /**
     * Saves the ScoutCard into the database
     * @param database used for interacting with the SQLITE db
     * @return int id of the saved ScoutCard
     */
    public int save(Database database)
    {
        int id = -1;

        //try to open the DB if it is not open
        if(!database.isOpen())
            database.open();

        if(database.isOpen())
            id = (int) database.setScoutCard(this);

        //set the id if the save was successful
        if(id > 0)
            setId(id);

        return getId();
    }

    /**
     * Deletes the ScoutCard from the database
     * @param database used for interacting with the SQLITE db
     * @return boolean if successful
     */
    public boolean delete(Database database)
    {
        boolean successful = false;

        //try to open the DB if it is not open
        if(!database.isOpen()) database.open();

        if(database.isOpen())
        {
            successful = database.deleteScoutCard(this);

        }

        return successful;
    }

    //endregion
}

File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Classes/Tables/ScoutCard.kt
package com.alphadevelopmentsolutions.frcscout.Classes.Tables

import com.alphadevelopmentsolutions.frcscout.Classes.Database
import com.alphadevelopmentsolutions.frcscout.Enums.StartingPiece
import com.alphadevelopmentsolutions.frcscout.Enums.StartingPosition
import java.text.SimpleDateFormat
import java.util.*

class ScoutCard(
        var id: Int,
        var matchId: String,
        var teamId: Int,
        var eventId: String,
        var allianceColor: String,
        var completedBy: String,
        var preGameStartingLevel: Int,
        var preGameStartingPosition: StartingPosition,
        var preGameStartingPiece: StartingPiece,
        var autonomousExitHabitat: Boolean,
        var autonomousHatchPanelsPickedUp: Int,
        var autonomousHatchPanelsSecuredAttempts: Int,
        var autonomousHatchPanelsSecured: Int,
        var autonomousCargoPickedUp: Int,
        var autonomousCargoStoredAttempts: Int,
        var autonomousCargoStored: Int,
        var teleopHatchPanelsPickedUp: Int,
        var teleopHatchPanelsSecuredAttempts: Int,
        var teleopHatchPanelsSecured: Int,
        var teleopCargoPickedUp: Int,
        var teleopCargoStoredAttempts: Int,
        var teleopCargoStored: Int,
        var endGameReturnedToHabitat: Int,
        var endGameReturnedToHabitatAttempts: Int,
        var defenseRating: Int,
        var offenseRating: Int,
        var driveRating: Int,
        var notes: String,
        var completedDate: Date,
        var isDraft: Boolean) : Table(TABLE_NAME, COLUMN_NAME_ID, CREATE_TABLE)
{
    companion object
    {

        val TABLE_NAME = "scout_cards"
        val COLUMN_NAME_ID = "Id"
        val COLUMN_NAME_MATCH_ID = "MatchId"
        val COLUMN_NAME_TEAM_ID = "TeamId"
        val COLUMN_NAME_EVENT_ID = "EventId"
        val COLUMN_NAME_ALLIANCE_COLOR = "AllianceColor"
        val COLUMN_NAME_COMPLETED_BY = "CompletedBy"

        val COLUMN_NAME_PRE_GAME_STARTING_LEVEL = "PreGameStartingLevel"
        val COLUMN_NAME_PRE_GAME_STARTING_POSITION = "PreGameStartingPosition"
        val COLUMN_NAME_PRE_GAME_STARTING_PIECE = "PreGameStartingPiece"

        val COLUMN_NAME_AUTONOMOUS_EXIT_HABITAT = "AutonomousExitHabitat"
        val COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_PICKED_UP = "AutonomousHatchPanelsPickedUp"
        val COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_SECURED = "AutonomousHatchPanelsSecured"
        val COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_SECURED_ATTEMPTS = "AutonomousHatchPanelsSecuredAttempts"
        val COLUMN_NAME_AUTONOMOUS_CARGO_PICKED_UP = "AutonomousCargoPickedUp"
        val COLUMN_NAME_AUTONOMOUS_CARGO_STORED = "AutonomousCargoStored"
        val COLUMN_NAME_AUTONOMOUS_CARGO_STORED_ATTEMPTS = "AutonomousCargoStoredAttempts"

        val COLUMN_NAME_TELEOP_HATCH_PANELS_PICKED_UP = "TeleopHatchPanelsPickedUp"
        val COLUMN_NAME_TELEOP_HATCH_PANELS_SECURED = "TeleopHatchPanelsSecured"
        val COLUMN_NAME_TELEOP_HATCH_PANELS_SECURED_ATTEMPTS = "TeleopHatchPanelsSecuredAttempts"
        val COLUMN_NAME_TELEOP_CARGO_PICKED_UP = "TeleopCargoPickedUp"
        val COLUMN_NAME_TELEOP_CARGO_STORED = "TeleopCargoStored"
        val COLUMN_NAME_TELEOP_CARGO_STORED_ATTEMPTS = "TeleopCargoStoredAttempts"

        val COLUMN_NAME_END_GAME_RETURNED_TO_HABITAT = "EndGameReturnedToHabitat"
        val COLUMN_NAME_END_GAME_RETURNED_TO_HABITAT_ATTEMPTS = "EndGameReturnedToHabitatAttempts"

        val COLUMN_NAME_DEFENSE_RATING = "DefenseRating"
        val COLUMN_NAME_OFFENSE_RATING = "OffenseRating"
        val COLUMN_NAME_DRIVE_RATING = "DriveRating"
        val COLUMN_NAME_NOTES = "Notes"

        val COLUMN_NAME_COMPLETED_DATE = "CompletedDate"
        val COLUMN_NAME_IS_DRAFT = "IsDraft"

        val CREATE_TABLE = "CREATE TABLE " + TABLE_NAME + " (" +
                COLUMN_NAME_ID + " INTEGER PRIMARY KEY," +
                COLUMN_NAME_MATCH_ID + " TEXT," +
                COLUMN_NAME_TEAM_ID + " INTEGER," +
                COLUMN_NAME_EVENT_ID + " TEXT," +
                COLUMN_NAME_ALLIANCE_COLOR + " TEXT," +
                COLUMN_NAME_COMPLETED_BY + " TEXT," +

                COLUMN_NAME_PRE_GAME_STARTING_LEVEL + " INTEGER," +
                COLUMN_NAME_PRE_GAME_STARTING_POSITION + " TEXT," +
                COLUMN_NAME_PRE_GAME_STARTING_PIECE + " TEXT," +

                COLUMN_NAME_AUTONOMOUS_EXIT_HABITAT + " INTEGER," +
                COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_PICKED_UP + " INTEGER," +
                COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_SECURED_ATTEMPTS + " INTEGER," +
                COLUMN_NAME_AUTONOMOUS_HATCH_PANELS_SECURED + " INTEGER," +
                COLUMN_NAME_AUTONOMOUS_CARGO_PICKED_UP + " INTEGER," +
                COLUMN_NAME_AUTONOMOUS_CARGO_STORED_ATTEMPTS + " INTEGER," +
                COLUMN_NAME_AUTONOMOUS_CARGO_STORED + " INTEGER," +

                COLUMN_NAME_TELEOP_HATCH_PANELS_PICKED_UP + " INTEGER," +
                COLUMN_NAME_TELEOP_HATCH_PANELS_SECURED_ATTEMPTS + " INTEGER," +
                COLUMN_NAME_TELEOP_HATCH_PANELS_SECURED + " INTEGER," +
                COLUMN_NAME_TELEOP_CARGO_PICKED_UP + " INTEGER," +
                COLUMN_NAME_TELEOP_CARGO_STORED_ATTEMPTS + " INTEGER," +
                COLUMN_NAME_TELEOP_CARGO_STORED + " INTEGER," +

                COLUMN_NAME_END_GAME_RETURNED_TO_HABITAT + " INTEGER," +
                COLUMN_NAME_END_GAME_RETURNED_TO_HABITAT_ATTEMPTS + " INTEGER," +

                COLUMN_NAME_DEFENSE_RATING + " INTEGER," +
                COLUMN_NAME_OFFENSE_RATING + " INTEGER," +
                COLUMN_NAME_DRIVE_RATING + " INTEGER," +
                COLUMN_NAME_NOTES + " TEXT," +

                COLUMN_NAME_COMPLETED_DATE + " INTEGER," +
                COLUMN_NAME_IS_DRAFT + " INTEGER)"

        /**
         * Clears all data from the classes table
         * @param database used to clear table
         * @param clearDrafts boolean if you want to include drafts in the clear
         */
        fun clearTable(database: Database, clearDrafts: Boolean)
        {
            clearTable(database, TABLE_NAME, clearDrafts)
        }

        /**
         * Returns arraylist of scout cards with specified filters from database
         * @param event if specified, filters scout cards by event id
         * @param match if specified, filters scout cards by match id
         * @param team if specified, filters scout cards by team id
         * @param scoutCard if specified, filters scout cards by scout card id
         * @param onlyDrafts if true, filters scout cards by draft
         * @param database used to load scout cards
         * @return arraylist of scout cards
         */
        fun getObjects(event: Event?, match: Match?, team: Team?, scoutCard: ScoutCard?, onlyDrafts: Boolean, database: Database): ArrayList<ScoutCard>?
        {
            return database.getScoutCards(event, match, team, scoutCard, onlyDrafts)
        }
    }

    /**
     * Gets the completed date formated for MySQL timestamp
     * @return MySQL time stamp formatted date
     */
    val completedDateForSQL: String
        get()
        {
            val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd H:mm:ss")

            return simpleDateFormat.format(completedDate)
        }


    override fun toString(): String
    {
        return "Team $teamId - Scout Card"
    }


    //endregion

    //region Load, Save & Delete

    /**
     * Loads the ScoutCard from the database and populates all values
     * @param database used for interacting with the SQLITE db
     * @return boolean if successful
     */
    override fun load(database: Database): Boolean
    {
        //try to open the DB if it is not open
        if (!database.isOpen) database.open()

        if (database.isOpen)
        {
            val scoutCards = getObjects(null, null, null, this, false, database)
            val scoutCard = if (scoutCards!!.size > 0) scoutCards[0] else null

            if (scoutCard != null)
            {
                matchId = scoutCard.matchId
                teamId = scoutCard.teamId
                eventId = scoutCard.eventId
                allianceColor = scoutCard.allianceColor
                completedBy = scoutCard.completedBy

                preGameStartingLevel = scoutCard.preGameStartingLevel
                preGameStartingPosition = scoutCard.preGameStartingPosition
                preGameStartingPiece = scoutCard.preGameStartingPiece

                autonomousExitHabitat = scoutCard.autonomousExitHabitat
                autonomousHatchPanelsPickedUp = scoutCard.autonomousHatchPanelsPickedUp
                autonomousHatchPanelsSecuredAttempts = scoutCard.autonomousHatchPanelsSecuredAttempts
                autonomousHatchPanelsSecured = scoutCard.autonomousHatchPanelsSecured
                autonomousCargoPickedUp = scoutCard.autonomousCargoPickedUp
                autonomousCargoStoredAttempts = scoutCard.autonomousCargoStoredAttempts
                autonomousCargoStored = scoutCard.autonomousCargoStored

                teleopHatchPanelsPickedUp = scoutCard.teleopHatchPanelsPickedUp
                teleopHatchPanelsSecuredAttempts = scoutCard.teleopHatchPanelsSecuredAttempts
                teleopHatchPanelsSecured = scoutCard.teleopHatchPanelsSecured
                teleopCargoPickedUp = scoutCard.teleopCargoPickedUp
                teleopCargoStoredAttempts = scoutCard.teleopCargoStoredAttempts
                teleopCargoStored = scoutCard.teleopCargoStored

                endGameReturnedToHabitat = scoutCard.endGameReturnedToHabitat
                endGameReturnedToHabitatAttempts = scoutCard.endGameReturnedToHabitatAttempts

                defenseRating = scoutCard.defenseRating
                offenseRating = scoutCard.offenseRating
                driveRating = scoutCard.driveRating
                notes = scoutCard.notes
                completedDate = scoutCard.completedDate
                isDraft = scoutCard.isDraft
                return true
            }
        }

        return false
    }

    /**
     * Saves the ScoutCard into the database
     * @param database used for interacting with the SQLITE db
     * @return int id of the saved ScoutCard
     */
    override fun save(database: Database): Int
    {
        var id = -1

        //try to open the DB if it is not open
        if (!database.isOpen)
            database.open()

        if (database.isOpen)
            id = database.setScoutCard(this).toInt()

        //set the id if the save was successful
        if (id > 0)
            id = id

        return id
    }

    /**
     * Deletes the ScoutCard from the database
     * @param database used for interacting with the SQLITE db
     * @return boolean if successful
     */
    override fun delete(database: Database): Boolean
    {
        var successful = false

        //try to open the DB if it is not open
        if (!database.isOpen) database.open()

        if (database.isOpen)
        {
            successful = database.deleteScoutCard(this)

        }

        return successful
    }

    //endregion
}
